processor/tracesamplerprocessor/tracesamplerprocessor.go

```diff
@@ -0,0 +1,168 @@
+	// Values greater or equal 100 are treated as "sample all traces".
+	SamplingPercentage float32
+	// HashSeed allows one to configure the hashing seed. This is important in scenarios where multiple layers of collectors
+	// have different sampling rates: if they use the same seed all passing one layer may pass the other even if they have
+	// different sampling rates, configuring different seeds avoids that.
+	HashSeed uint32
+}
+
+// InitFromViper updates TraceSamplerCfg according to the viper configuration.
+func (tsc *TraceSamplerCfg) InitFromViper(v *viper.Viper) (*TraceSamplerCfg, error) {
+	if v == nil {
+		return nil, errors.New("v is nil")
+	}
+	if err := v.UnmarshalKey(samplingPercentageCfgTag, &tsc.SamplingPercentage); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal %q: %v", samplingPercentageCfgTag, err)
+	}
+	if err := v.UnmarshalKey(hashSeedCfgTag, &tsc.HashSeed); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal %q: %v", hashSeedCfgTag, err)
+	}
+	return tsc, nil
+}
+
+type tracesamplerprocessor struct {
+	nextConsumer       consumer.TraceConsumer
+	scaledSamplingRate uint32
+	hashSeed           uint32
+}
+
+var _ processor.TraceProcessor = (*tracesamplerprocessor)(nil)
+
+// NewTraceProcessor returns a processor.TraceProcessor that will perform head sampling according to the given
+// configuration.
+func NewTraceProcessor(nextConsumer consumer.TraceConsumer, cfg TraceSamplerCfg) (processor.TraceProcessor, error) {
+	if nextConsumer == nil {
+		return nil, errors.New("nextConsumer is nil")
+	}
+
+	return &tracesamplerprocessor{
+		nextConsumer: nextConsumer,
+		// Adjust sampling percentage on private so recalculations are avoided.
+		scaledSamplingRate: uint32(cfg.SamplingPercentage * percentageScaleFactor),
+		hashSeed:           cfg.HashSeed,
+	}, nil
+}
+
+func (tsp *tracesamplerprocessor) ConsumeTraceData(ctx context.Context, td data.TraceData) error {
+	scaledSamplingRate := tsp.scaledSamplingRate
+	if scaledSamplingRate >= numHashBuckets {
+		return tsp.nextConsumer.ConsumeTraceData(ctx, td)
+	}
+
+	sampledTraceData := data.TraceData{
+		Node:         td.Node,
+		Resource:     td.Resource,
+		SourceFormat: td.SourceFormat,
+	}
+
+	sampledSpans := make([]*tracepb.Span, 0, len(td.Spans))
+	for _, span := range td.Spans {
+		// If one assumes random trace ids hashing may seems avoidable, however, traces can be coming from sources
+		// with various different criterias to generate trace id and perhaps were already sampled without hashing.
+		// Hashing here prevents bias due to such systems.
+		if hash(span.TraceId, tsp.hashSeed)&bitMaskHashBuckets < scaledSamplingRate {
+			sampledSpans = append(sampledSpans, span)
+		}
+	}
+
+	sampledTraceData.Spans = sampledSpans
+
+	return tsp.nextConsumer.ConsumeTraceData(ctx, sampledTraceData)
+}
+
+// hash is a murmur3 hash function, see http://en.wikipedia.org/wiki/MurmurHash.
+func hash(key []byte, seed uint32) (hash uint32) {
```

Comment:
Is there a reason not to use a library that implements this? https://github.com/spaolacci/murmur3

Reply:
Avoid a new dependency for a standard like murmur3 that has a very small implementation.

Reply:
ok